cmd/load-generator/state.go

```diff
@@ -0,0 +1,472 @@
+	apiBase         string
+	domainBase      string
+	email           string
+	maxRegs         int
+	maxNamesPerCert int
+	realIP          string
+	certKey         *ecdsa.PrivateKey
+
+	operations []func(*State, *context) error
+
+	rMu  sync.RWMutex
+	regs []*registration
+
+	challSrv    *challSrv
+	callLatency *latencyFile
+	client      *http.Client
+
+	getTotal  int64
+	postTotal int64
+	respCodes map[int]*respCode
+	cMu       sync.Mutex
+
+	wg *sync.WaitGroup
+}
+
+type rawRegistration struct {
+	Certs          []string `json:"certs"`
+	FinalizedAuthz []string `json:"finalizedAuthz"`
+	RawKey         []byte   `json:"rawKey"`
+}
+
+type snapshot struct {
+	Registrations []rawRegistration
+}
+
+func (s *State) numRegs() int {
+	s.rMu.RLock()
+	defer s.rMu.RUnlock()
+	return len(s.regs)
+}
+
+// Snapshot will save out generated registrations and certs (ignoring authorizations)
+func (s *State) Snapshot(filename string) error {
+	fmt.Printf("[+] Saving registrations to %s\n", filename)
+	snap := snapshot{}
+	// assume rMu lock operations aren't happening right now
+	for _, reg := range s.regs {
+		k, err := x509.MarshalECPrivateKey(reg.key)
+		if err != nil {
+			return err
+		}
+		snap.Registrations = append(snap.Registrations, rawRegistration{
+			Certs:          reg.certs,
+			FinalizedAuthz: reg.finalizedAuthz,
+			// RawKey:         x509.MarshalPKCS1PrivateKey(reg.key),
+			RawKey: k,
+		})
+	}
+	cont, err := json.Marshal(snap)
+	if err != nil {
+		return err
+	}
+	return ioutil.WriteFile(filename, cont, os.ModePerm)
+}
+
+// Restore previously generated registrations and certs
+func (s *State) Restore(filename string) error {
+	fmt.Printf("[+] Loading registrations from %s\n", filename)
+	content, err := ioutil.ReadFile(filename)
+	if err != nil {
+		return err
+	}
+	snap := snapshot{}
+	err = json.Unmarshal(content, &snap)
+	if err != nil {
+		return err
+	}
+	for _, r := range snap.Registrations {
+		// key, err := x509.ParsePKCS1PrivateKey(r.RawKey)
+		key, err := x509.ParseECPrivateKey(r.RawKey)
+		if err != nil {
+			continue
+		}
+		// key.Precompute()
+		signer, err := jose.NewSigner(jose.RS256, key)
+		if err != nil {
+			continue
+		}
+		s.regs = append(s.regs, &registration{
+			key:    key,
+			signer: signer,
+			certs:  r.Certs,
+			//			auths:  r.Auths,
+		})
+	}
+	return nil
+}
+
+// New returns a pointer to a new State struct or an error
+func New(apiBase string, keySize int, domainBase string, realIP string, maxRegs int, latencyPath string, userEmail string, operations []string) (*State, error) {
+	certKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		return nil, err
+	}
+	client := &http.Client{
+		Transport: &http.Transport{
+			DialContext: (&net.Dialer{
+				Timeout:   10 * time.Second,
+				KeepAlive: 30 * time.Second,
+			}).DialContext,
+			TLSHandshakeTimeout: 5 * time.Second,
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: true, // CDN bypass can cause validation failures
```

Comment:
The bypass is because the internal TLS server (that we access when bypassing the CDN) serves an certificate with a non-public root. I don't _think_ there is a simple way to test this by inspecting responses etc, but I'd be happy to move it to a config bool or something?